src/routes/notifications/notifications.controller.spec.ts

```diff
@@ -93,29 +109,47 @@ describe('Notifications Controller (Unit)', () => {
         .with('safeRegistration', safeRegistrations)
         .build();
       mockNetworkService.get.mockImplementation((url) => {
-        const getChainUrlPattern = `${safeConfigUrl}/api/v1/chains/`;
-        if (url.includes(getChainUrlPattern)) {
-          return Promise.resolve({ data: sample(chains) });
+        const targetChain = chains.find(
+          (chain) => url === `${safeConfigUrl}/api/v1/chains/${chain.chainId}`,
+        );
+        if (targetChain) {
+          return Promise.resolve({ data: targetChain });
         }
         return Promise.reject(`No matching rule for url: ${url}`);
       });
-      // return a client error for just one provider
-      const error = new NetworkResponseError(
-        faker.datatype.number({ min: 400, max: 499 }),
-      );
-      mockNetworkService.post
-        .mockImplementationOnce(() => Promise.reject(error))
-        .mockImplementation(() => Promise.resolve());
+      // return a client error for two chains
+      const errorChainIds = [chains[1].chainId, chains[2].chainId];
+      mockNetworkService.post.mockImplementation((url) => {
+        const targetChain = chains.find(
+          (chain) =>
+            url === `${chain.transactionService}/api/v1/notifications/devices/`,
+        );
+        if (!targetChain) {
+          return Promise.reject(`No matching rule for url: ${url}`);
+        }
+        return errorChainIds.includes(targetChain?.chainId)
+          ? Promise.reject(
+              new NetworkResponseError(
+                faker.datatype.number({ min: 400, max: 499 }),
+              ),
+            )
+          : Promise.resolve();
+      });
 
       await request(app.getHttpServer())
         .post('/register/notifications')
         .send(registerDeviceDto)
-        .expect(400);
+        .expect(400)
+        .then(({ body }) =>
+          expect(body).toMatchObject({
+            message: `Push notification registration failed for chain IDs: ${errorChainIds}`,
+          }),
+        );
     });
 
     it('Server errors returned from provider', async () => {
-      const chains: Chain[] = range(4).map((idx) =>
-        chainBuilder().with('chainId', idx.toString()).build(),
+      const chains: Chain[] = range(21, 30).map((chainId) =>
```

Comment:
OOC: Why do we increase the range on each test?

Reply:
The problem I've been facing with this is that `TransactionApiManager` is not stateless, because it has this internal state:
```javascript
private transactionApiMap: Record<string, TransactionApi> = {};
```
So not every test having N chains will use `mockNetworkService.get` N times, because if the internal state of `TransactionApiManager` already contains that chain, the mock function doesn't get called. So this makes the value of `chain.transactionService` not controlled and therefore the URL for the subsequent POST call wouldn't match.

This problem is a bit tricky to bypass and that's something that it's overcomplicating the tests as you noted on [this comment](https://github.com/5afe/safe-client-gateway-nest/pull/267#discussion_r1098651611). I see some mitigations but I'm open to suggestions:

- Move the internal state of `TransactionApiManager` from memory to redis: this solution could be useful but is out of the scope of this PR I guess.
- Implement a function to clear the internal state of `TransactionApiManager` and call that function on `beforeEach` hook: not a fan of this one since we would be changing production code due to testing implementation requirements.
- Remove the numerical ranges and use uuids for `chainIds` in these tests: probably the easier but tests will remain verbose and pretty difficult to read and follow.
- Make an exception and rely on patterns instead of strict URL matching when implementing mock functions: I think we could go this way.

What do you think @fmrsabino?

Reply:
I'll give this a second thought and I'll try to come up with a simpler solution @fmrsabino.

Reply:
But the `beforeEach` should give you a clean state of the application or that is not happening?

Reply:
Ok, I found the reason, the app was correctly initialized on each test run, the problem I had was the test wasn't clearing the cache, I forgot to include:
```javascript
fakeCacheService.clear();
```
in the `beforeEach` hook 🤦🏻‍♂️

Now the previous implementation with the strict URL matching and without the need to use different ranges for `chainId` in each test works. But I'm pretty happy with the current implementation using URL patterns TBH. What do you think?

Reply:
> Ok, I found the reason, the app was correctly initialized on each test run, the problem I had was the test wasn't clearing the cache, I forgot to include:
> 
> ```js
> fakeCacheService.clear();
> ```
> 
> in the `beforeEach` hook 🤦🏻‍♂️

That's strange because, as you mentioned before, if the `TransactionApiManager` was the root cause of the issue, it doesn't use Redis. It's just a local `Record`.

The `fakeCacheService` replaces Redis (as a fake implementation).

> Now the previous implementation with the strict URL matching and without the need to use different ranges for `chainId` in each test works. But I'm pretty happy with the current implementation using URL patterns TBH. What do you think?

I think it's a nice improvement and the tests are more readable in my opinion! In this case it works – for some other tests it could also work but some of them test that the path is called with some specific parameters (safe address, transaction hash, etc.) – there's also value in knowing and having the test fail if the URL does not match.

Reply:
> That's strange because, as you mentioned before, if the `TransactionApiManager` was the root cause of the issue, it doesn't use Redis. It's just a local `Record`.

I assumed the problem was in `TransactionApiManager` but it wasn't, my fault, the problem was in the deeper layer because the cache was not being cleared 😄

Reply:
> > That's strange because, as you mentioned before, if the `TransactionApiManager` was the root cause of the issue, it doesn't use Redis. It's just a local `Record`.
> 
> I assumed the problem was in `TransactionApiManager` but it wasn't, my fault, the problem was in the deeper layer because the cache was not being cleared 😄

Got it 🙌